aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs

```diff
@@ -138,101 +205,154 @@ pub(crate) fn is_valid_txn_arg<S: MoveResolverExt>(
     }
 }
 
-// Validate arguments. Walk through the arguments and according to the signature
-// validate arguments that require so.
+// Construct arguments. Walk through the arguments and according to the signature
+// construct arguments that require so.
 // TODO: This needs a more solid story and a tighter integration with the VM.
-// Validation at the moment is only for Strings and Vector of them, so we
-// manually walk the serialized args until we find a string.
-// This is obviously brittle and something to change at some point soon.
-pub(crate) fn validate_args<S: MoveResolverExt>(
-    session: &SessionExt<S>,
+pub(crate) fn construct_args<S: MoveResolverExt>(
+    session: &mut SessionExt<S>,
     idxs: &[usize],
-    args: &[Vec<u8>],
+    args: &mut [Vec<u8>],
     func: &LoadedFunctionInstantiation,
+    allowed_structs: &ConstructorMap,
 ) -> Result<(), VMStatus> {
-    for (idx, (ty, arg)) in func.parameters.iter().zip(args.iter()).enumerate() {
+    // Perhaps in a future we should do proper gas metering here
+    let mut gas_meter = UnmeteredGasMeter;
+    for (idx, ty) in func.parameters.iter().enumerate() {
         if !idxs.contains(&idx) {
             continue;
         }
-        let arg_len = arg.len();
+        let arg = &mut args[idx];
         let mut cursor = Cursor::new(&arg[..]);
-        validate_arg(session, ty, &mut cursor, arg_len)?;
+        let mut new_arg = vec![];
+        recursively_construct_arg(
+            session,
+            &ty.subst(&func.type_arguments).unwrap(),
+            allowed_structs,
+            &mut cursor,
+            &mut gas_meter,
+            &mut new_arg,
+        )?;
+        // Check cursor has parsed everything
+        // Unfortunately, is_empty is only enabled in nightly, so we check this way.
+        if cursor.position() != arg.len() as u64 {
+            return Err(VMStatus::Error(
+                StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT,
```

Comment:
Shouldn't this be NUMBER_OF_ARGUMENTS_MISMATCH? Let's also add an error message here instead of None